plugins/manager/manager.go

```diff
@@ -0,0 +1,242 @@
+		pluginInstances: make(map[plugins.PluginID]PluginInstance),
+		plugins:         make(map[plugins.PluginID]*pluginInfo),
+	}
+}
+
+// Load is responsible for registering and executing the plugins configured for
+// use by the Autoscaler agent.
+func (pm *PluginManager) Load() error {
+
+	for t, cfgs := range pm.cfg {
+		for _, cfg := range cfgs {
+
+			// Figure out if the plugin is internal or external, then perform
+			// the loading of the config into the manager store.
+			if pm.useInternal(cfg.Driver) {
+				pm.loadInternalPlugin(cfg, t)
+			} else {
+				pm.loadExternalPlugin(cfg, t)
+			}
+		}
+	}
+
+	return pm.dispensePlugins()
+}
+
+// KillPlugins calls Kill on all plugins currently dispensed.
+func (pm *PluginManager) KillPlugins() {
+	for id, v := range pm.pluginInstances {
+		pm.logger.Info("shutting down plugin", "plugin_name", id.Name)
+		v.Kill()
+	}
+}
+
+// Dispense returns a PluginInstance for use by safely obtaining the
+// PluginInstance from storage if we have it.
+func (pm *PluginManager) Dispense(name, pluginType string) (PluginInstance, error) {
+	pm.pluginInstancesLock.RLock()
+	defer pm.pluginInstancesLock.RUnlock()
+
+	// Attempt to pull our plugin instance from the store and pass this to the
+	// caller.
+	//
+	// TODO(jrasell) if we do not find the instance, we should probably try and
+	//  dispense the plugin. We should also check the plugin instance has not
+	//  exited.
+	inst, ok := pm.pluginInstances[plugins.PluginID{Name: name, PluginType: pluginType}]
+	if !ok {
+		return nil, fmt.Errorf("failed to dispense plugin: %s of type %s is not stored", name, pluginType)
+	}
+	return inst, nil
+}
+
+// dispensePlugins launches all configured plugins. It is responsible for
+// executing external binaries as well as setting the config on all plugins so
+// they are in a ready state. Any errors from this process will result in the
+// agent failing on startup.
+func (pm *PluginManager) dispensePlugins() error {
+
+	var mErr multierror.Error
+
+	pm.pluginsLock.RLock()
+	p := pm.plugins
+	pm.pluginsLock.RUnlock()
+
+	for pID, pInfo := range p {
+
+		var (
+			inst PluginInstance
+			err  error
+		)
+		if pInfo.factory != nil {
+			inst, err = pm.launchInternalPlugin(pID, pInfo)
+		} else {
+			inst, err = pm.launchExternalPlugin(pID, pInfo)
+		}
+
+		// If we got an error dispensing the plugin, add this to the muilterror
+		// and continue the loop.
+		if err != nil {
+			_ = multierror.Append(&mErr, fmt.Errorf("failed to dispense plugin %s: %v", pID.Name, err))
+			continue
+		}
+
+		// Perform the SetConfig on the plugin to ensure its state is as the
+		// operator desires.
+		if err := inst.Plugin().(base.Plugin).SetConfig(pInfo.config); err != nil {
+			inst.Kill()
+			_ = multierror.Append(&mErr, fmt.Errorf("failed to set config on plugin %s: %v", pID.Name, err))
+			continue
+		}
+
+		// Store our plugin instance.
+		pm.pluginInstancesLock.Lock()
+		pm.pluginInstances[pID] = inst
+		pm.pluginInstancesLock.Unlock()
```

Comment:
we only call this method once, so it's not necessary, but.... should we explicitly clear out the instance map at the top of the method?

also, not sure whether the repeated locking is necessary (i don't think there's any concurrent access to this struct right now); is this in preparation for concurrent loading?

Reply:
Yeah I we can clear it out. What is the advantage of doing this seeing as the method is only called once? To protect in the future if it gets called > 1 time?

Currently its locking for the sake of locking; but in the future this will be needed when the manager makes attempts to re-launch and dispense plugins which have failed after normal startup.